lib/nethealth/nethealth.go

```diff
@@ -0,0 +1,587 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	s.client, err = kubernetes.NewForConfig(config)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.conn, err = icmp.ListenPacket("ip4:icmp", "0.0.0.0")
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.clock = clockwork.NewRealClock()
+	go s.loop()
+	go s.loopServiceDiscovery()
+	go s.serve()
+
+	mux := http.ServeMux{}
+	mux.Handle("/metrics", promhttp.Handler())
+	s.httpServer = &http.Server{Addr: fmt.Sprint(":", s.config.PrometheusPort), Handler: &mux}
+	go func() {
+		if err := s.httpServer.ListenAndServe(); err != http.ErrServerClosed {
+			s.Fatalf("ListenAndServe(): %s", err)
+		}
+	}()
+
+	s.Info("Started nethealth with config:")
+	s.Info("  PrometheusPort: ", s.config.PrometheusPort)
+	s.Info("  Namespace: ", s.config.Namespace)
+	s.Info("  NodeName: ", s.config.NodeName)
+	s.Info("  Selector: ", s.selector)
+	s.Info("  ServiceDiscoveryQuery: ", s.config.ServiceDiscoveryQuery)
+
+	return nil
+}
+
+// loop is the main processing loop for sending/receiving heartbeats.
+func (s *Server) loop() {
+	heartbeatTicker := s.clock.NewTicker(heartbeatInterval)
+	defer heartbeatTicker.Stop()
+
+	resyncTicker := s.clock.NewTicker(resyncInterval)
+	defer resyncTicker.Stop()
+
+	for {
+		select {
+		//
+		// Re-sync cluster peers
+		//
+		case <-resyncTicker.Chan():
+			err := s.resyncPeerList()
+			if err != nil {
+				s.WithError(err).Error("Unexpected error re-syncing the list of peer nodes.")
+			}
+
+			err = s.resyncNethealthPods()
+			if err != nil {
+				s.WithError(err).Error("Unexpected error re-syncing the list of peer pods.")
+			}
+		case <-s.triggerResync:
+			err := s.resyncPeerList()
+			if err != nil {
+				s.WithError(err).Error("Unexpected error re-syncing the list of peer nodes.")
+			}
+
+			err = s.resyncNethealthPods()
+			if err != nil {
+				s.WithError(err).Error("Unexpected error re-syncing the list of peer pods.")
+			}
+
+		//
+		// Send a heartbeat to each peer we know about
+		// Check for peers that are timing out / down
+		//
+		case <-heartbeatTicker.Chan():
+			s.checkTimeouts()
+			for _, peer := range s.peers {
+				s.sendHeartbeat(peer)
+			}
+
+		//
+		// Rx heartbeats responses from peers
+		//
+		case rx := <-s.rxMessage:
+			err := s.processAck(rx)
+			if err != nil {
+				s.WithFields(logrus.Fields{
+					logrus.ErrorKey: err,
+					"peer_addr":     rx.peerAddr,
+					"rx_time":       rx.rxTime,
+					"message":       rx.message,
+				}).Error("Error processing icmp message.")
+			}
+		}
+	}
+}
+
+// loopServiceDiscovery uses cluster-dns service discovery as a lightweight check for pod changes
+// and will trigger a resync if the cluster DNS service discovery changes
+func (s *Server) loopServiceDiscovery() {
+	s.Info("Starting DNS service discovery for nethealth pod.")
+	ticker := s.clock.NewTicker(dnsDiscoveryInterval)
+	query := s.config.ServiceDiscoveryQuery
+
+	previousNames := []string{}
+
+	for {
+		<-ticker.Chan()
```

Comment:
I don't believe that's correct. Golang routines don't block program exit.

Reply:
Ah, so you don't need to wait for goroutines to exit and clean up?

Reply:
Yea, I don't really see a benefit.

It's one of those things, where I had originally implemented a Stop function, but actually realized I had implemented it wrong and it was non effective. So it does tie the library code to the way it's used, because the nethealth pings can't be shutdown without process exit, but this seems like a reasonable tradeoff vs adding complexity that doesn't necessarily work right. This ties into other philosophies I'm personally a fan of like crash only software, so just exit the process as if it were to crash.

So that's why I was thinking about it a bit, and in this case I don't really see an important reason to not just let the process exit immediately, and allow the kernel to release the resources as if the process crashed.